### PR TITLE
[Distributed] Remove class-bound restriction from target invocation decoder

### PR DESCRIPTION
**Pull request by xedin**
- Remove artificial restriction from `DistributedTargetInvocationDecoder` to be class-bound;
- Make `executedDistributedTarget` and accessor functions generic over type of decoder.

Resolves: rdar://problem/88719791

### CODE REVIEW

**Comment by xedin** [comment]
@ktoso We need to add tests for structs and enums but I didn't want to just copy-paste encoders/decoders or clone _remoteCall.swift per nominal type, so ideas are welcome!

**Comment by xedin** [comment]
The annoying limitation here is that we cannot have multiple fake systems in a different module either because symbol for `decodeNextArgument` from a different module need to get requested during SILGen which is not yet implemented...

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
> The annoying limitation here is that we cannot have multiple fake systems in a different module either because symbol for decodeNextArgument from a different module need to get requested during SILGen which is not yet implemented...

I think I can give this a shot right now

File: stdlib/public/Distributed/DistributedActor.cpp [hunk 1]
```diff
@@ -40,11 +40,11 @@ void *swift_distributed_getGenericEnvironment(const char *targetNameStart,
   return accessor ? accessor->GenericEnvironment.get() : nullptr;
 }
 
-/// func _executeDistributedTarget(
+/// func _executeDistributedTarget<D: DistributedTargetInvocationDecoder>(
 ///    on: AnyObject,
 ///    _ targetName: UnsafePointer<UInt8>,
 ///    _ targetNameLength: UInt,
-///    argumentDecoder: AnyObject,
+///    argumentDecoder: inout D,
```
Comment (ktoso) [inline stdlib/public/Distributed/DistributedActor.cpp:47]: 🥳

File: stdlib/public/Distributed/DistributedActor.cpp [hunk 2]
```diff
@@ -75,14 +77,18 @@ TargetExecutorSignature::FunctionType swift_distributed_execute_target;
 ///   - a list of witness tables
 ///   - a number of witness tables in the buffer
 ///   - a reference to an actor to execute method on.
+///   - a type of the argument decoder
+///   - a witness table associated with argument decoder value
```
Comment (ktoso) [inline stdlib/public/Distributed/DistributedActor.cpp:81]: i see, so metadata and the table gets passed... ok :)

**Review by ktoso — APPROVED** [review]
Thanks a lot -- will think of a way to test more cases and look into the retaining the decodeNextArgument call now

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by A3ew3** [comment]
xedin:dist-generic-accessors
Sources/Basics/HTPClient+URLSession.swift YES
Do Not Allow Advertiser Tracking

Objective-CSwiftUnity SDK

// Set AdvertiserTrackingEnabled to NO if a device does not provide consent [FBSDKSettings setAdvertiserTrackingEnabled:NO];

**Comment by xedin** [comment]
@swift-ci please test Windows platform

**Comment by A3ew3** [comment]
`[AdvertiserTrackingEnabled](;Sources/Build/BuildPlan.swift yes or Do Not Allow Advertiser Tracking  Objective-CSwiftUnity SDK  // Set AdvertiserTrackingEnabled to NO if a device does not provide consent [FBSDKSettings setAdvertiserTrackingEnabled:NO];) `

**Comment by A3ew3** [comment]
gh pr checkout 41415``

File: lib/IRGen/GenDistributed.cpp [hunk 3]
```diff
@@ -131,8 +141,8 @@ class DistributedAccessor {
   void emit();
 
 private:
-  void decodeArguments(llvm::Value *decoder, llvm::Value *argumentTypes,
-                       Explosion &arguments);
+  void decodeArguments(const ArgumentDecoderInfo &decoder,
```
Comment (A3ew3) [inline lib/IRGen/GenDistributed.cpp:144]: Do Not Allow Advertiser Tracking

Objective-CSwiftUnity SDK

// Set AdvertiserTrackingEnabled yas or
 NO if a device does not provide consent
 [FBSDKSettings setAdvertiserTrackingEnabled:NO];

**Review by A3ew3 — CHANGES_REQUESTED** [review]
`

**Comment by ktoso** [comment]
Seems this helped, failure was lldb   `lldb-shell :: SwiftREPL/BreakpointSimple.test`

**Comment by ktoso** [comment]
@swift-ci please test macOS

**Pull request merged**